src/Compiler/pars.fsy

```diff
@@ -4430,6 +4430,36 @@ declExpr:
      { let trivia: SynExprYieldOrReturnTrivia = { YieldOrReturnKeyword = rhs parseState 1 }
        SynExpr.YieldOrReturn(($1, not $1), $2, (unionRanges (rhs parseState 1) $2.Range), trivia) }
 
+  | YIELD declExpr COLON typ
+     { let trivia: SynExprYieldOrReturnTrivia = { YieldOrReturnKeyword = rhs parseState 1 }
+       let typedExpr = SynExpr.Typed($2, $4, unionRanges $2.Range $4.Range)
+       parseState.LexBuffer.CheckLanguageFeatureAndRecover LanguageFeature.AllowTypedLetOrUseBang typedExpr.Range
+       SynExpr.YieldOrReturn(($1, not $1), typedExpr, (unionRanges (rhs parseState 1) $4.Range), trivia) }
+
+  | YIELD declExpr opt_topReturnTypeWithTypeConstraints
+     { let trivia: SynExprYieldOrReturnTrivia = { YieldOrReturnKeyword = rhs parseState 1 }
+       let expr = 
+           match $3 with
+           | None -> $2
+           | Some(_, SynReturnInfo((ty, _), m)) -> 
+               SynExpr.Typed($2, ty, unionRanges $2.Range m)
+       parseState.LexBuffer.CheckLanguageFeatureAndRecover LanguageFeature.AllowTypedLetOrUseBang expr.Range
+       SynExpr.YieldOrReturn(($1, not $1), expr, (unionRanges (rhs parseState 1) expr.Range), trivia) }
+
+  | YIELD_BANG declExpr COLON typ
```

Comment:
Code organisation - could we have all the `YIELD` rules first, and `YIELD_BANG` afterwards?